File: src/0199-binary-tree-right-side-view/solution.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        """
        We have to use BFS here to go down the tree level by level. The answer
        is to NOT just only go through the right side of the tree. We actually also need
        the left side of the tree if the right side of the tree is empty.
        Instead, we only add nodes to the result, if the node is the LAST PROCESSED
        node, or in this case the right-most node
        """
        q = deque()
        res = []
        level = 1
        if root:
            q.append(root)
        while q:
            print(f"The level is {level}")
            curr_length_of_queue = len(q)
            for i in range(len(q)):
                curr = q.popleft()
                if i == curr_length_of_queue - 1:
                    res.append(curr.val)
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            level += 1
        return res
        

File: src/0199-binary-tree-right-side-view/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
    /***
     * Basically the right side view node, is the node at the last item to be processed in the queue
     * of the BFS algo
     */
    const q = new Queue.Queue();
    const result = [];
    if (!root) {
        return result;
    }

    q.enqueue(root);
    while (!q.isEmpty()) {
        const lengthOfQueue = q.size();
        for (let i = 0; i < lengthOfQueue; i++) {
            const curr = q.dequeue();
            // If we are the last item to be processed, means most rightmost node
            if (i === lengthOfQueue - 1) {
                result.push(curr.val)
            }
            if (curr.left) {
                q.enqueue(curr.left);
            }

            if (curr.right) {
                q.enqueue(curr.right);
            }
        }
    }

    return result;
};
